BERT Inference on CPU (#100)
huggingface/blog: benchmark-2021-part1.md

### Core count scaling - Does using more cores actually improve performances?

At this stage, you may wonder what is the point of allocating only a subset of the cores to the task rather than throwing
all the horses to achieve the minimum latency. 

Rationally, throwing more resources to the task might give better results. 
This statement really depends on the problem size.   
Indeed, it's possible than for small problems putting more threads at work
doesn't bring improvements over the latency measurements.

In order to illustrate this, the figure below takes different problem sizes (`sequence length = {32, 128, 512}`)
and reports the latencies with respect to the number of threads used for running
computations for both PyTorch and TensorFlow.

Limiting the number of resources involved in computation is done by limiting the number of threads involved in 
**intra** operations (_**intra** here means inside an operator doing computation, also known as "kernel"_). 

This is achieved through the following API:
- PyTorch: `torch.set_num_threads(x)`
- TensorFlow: `tf.config.threading.set_intra_op_parallelism_threads(x)`

> ![pytorch_tensorflow_intraops_scaling](assets/19_benchmark_2021_part1/imgs/core_count_scaling.svg)

Reviewer: I find the title a bit misleading here. I was expecting the batch size to change along the way but it seems to be always 1. Also, I find it a bit hard to parse the various parameters.

> As you can see, depending on the problem size, the number of threads involved in the computations has a positive impact

Reviewer: I find the phrasing a bit weird here, what about something like this:

> As you can see, increasing the number of threads usually has a positive impact, but the size of the problem limits this impact a lot.
>
> For small-sized problems, using only 8 or 16 threads already gives the best performances. The same applies to medium-sized problems where using a single CPU (24 threads) gives the best results. For large-sized problems though, the overhead of the cross-socket communication is covered by the computations cost, thus benefiting from using all the cores available on the system.

on the latency measurements. 

Still, for small-sized problems, using only 8 or 16 threads already gives the best performances,
the same applies for medium-sized problems where using a single CPU (24 threads) gives the best results. Finally, for
large-sized problems, the overhead of the cross-socket communication is covered by the computations cost, thus benefiting from
using all the cores available on the system.

### Cores and Threads on Modern CPUs

On our way towards optimizing CPU inference for better usage of the CPU cores you might have already seen -_at least for the
past 20 years_- modern CPUs specifications reports "cores" and "threads" or "physical" and "logical" numbers. 
These notions refer to a mechanism called **Simultaneous Multi-Threading** (SMT).

To put things simple, imagine two tasks **A** and **B**, executing in parallel, each on its own thread.  
At some point, there is a high probability these two tasks will have to wait for some resources to be fetched from main memory, SSD, HDD 
or even the network.  
During these periods the core executing the task is in an **Idle** state waiting for the resources to arrive, and effectively doing nothing...

> Now, with **SMT**, the **two threads for task A and B** will be scheduled on the same **physical core**, 

Reviewer: Might be worth mentioning that `Hyper Threading` is the name Intel gave to `SMT`.

but their execution will be interleaved:  
When task **A** receives a system interrupt, task **B** will resume 
its execution, then task **A**, etc. which increases overall core utilization.

<!-- The figure below explains the above visually ( [source](https://appuals.com/how-does-hyper-threading-work-in-intel-core-i7-processors/) ) -->

<img class="centered" alt="Intel Hyper Threading technology" src="assets/19_benchmark_2021_part1/imgs/hyper_threading_explained.png" />

Back to our model inference workload... If you think about it, in a perfect world and very optimized setup, computations take the majority of time. 

In this context, using the logical cores shouldn't bring us any performance benefits because our computation work

> (*task A in the example above*) is not letting the second one (*task B*) being schedule on the CPU core.

Reviewer: ```suggestion
(*task A in the example above*) is not letting the second one (*task B*) being scheduled on the CPU core.
```

to the range `[0...47]`.  
Moreover, the range `[0...47]` spawn cores across both sockets, so we need to bind the memory allocations accordingly (`0,1`).

_Please note **this setup is significantly slower than just launching without `numactl`** for small-sized problem.  
This slowness is expected as the computations span over the two CPUs it involves cross-socket communication overhead 
which in this case is higher than the overall computations time._

## Multi-Stream Inference - End to end 

If you keep on reading until here, you should now be in a good shape to setup optimized inference workload on CPU.  
Now, we are going to highlight some possiblities offered by the beefy hardware we have and the tuning the knobs described before to scale 
our inference as linearly as possible.

Let's start simple, if we want to spawn 2 instances, one on each socket with 24 cores assigned:
```shell
numactl -C 0-23 -m 0 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=24
numactl -C 24-47 -m 1 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=24
```

> Starting from here, both instances doesn't share any resources between them and everything is operating as its maximum efficiency from a 

Reviewer: ```suggestion
Starting from here, both instances don't share any resources between them and everything is operating at its maximum efficiency from a 
```

numactl -C 0-7 -m 0 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=32 backend.name=pytorch backend.num_threads=8

# Medium-sized problems (32 > sequence >= 384) use remaining 16 cores (on CPU 0 - (8+16)/24 cores used)
numactl -C 8-23 -m 0 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=16

# Large sized problems (sequence >= 512) use the entire CPU on the second socket (on CPU 1 - 24/24 cores used)
numactl -C 24-37 -m 1 python3 src/main.py model=bert-base-cased batch_size=1 sequence_length=128 backend.name=pytorch backend.num_threads=24
```

The following section summarizes the performances of Multi-Stream Inference, leveraging all the knobs explained above.

### Batch size scaling results

The first chart below reports the best latency setup depending on the problem size (_w.r.t the sequence length_).
This corresponds to taking the **maximum** inference time for all the instance for a same problem size.
The second one reports the actually scaling efficiency by **summing** the throughput for all instances for a same problem size.

> ![Batch scaling experiment for PyTorch and Tensorflow](assets/19_benchmark_2021_part1/imgs/batch_scaling_exp.svg)

Reviewer: I find the parameters hard to read on this graph too.

![Batch scaling experiment for PyTorch and Tensorflow](assets/19_benchmark_2021_part1/imgs/batch_scaling_exp_throughput.svg)

## Conclusion

> Through this blog post we covered the basic BERT inference results for both PyTorch and TensorFlow one can expect without from a simple PyPi install and without any further tuning.

Reviewer: ```suggestion
Through this blog post, we covered the basic BERT inference results for both PyTorch and TensorFlow one can expect from a simple PyPi install and without any further tuning.
```

on socket 1 would ask the system to go over the sockets shared bus to exchange memory, thus leading to an undesired overhead._

### Tuning Process Affinity & Memory Allocation Policy

Now that we have all the knobs required to control the resources' allocation of our model instances we go further and see how to
effectively deploy those and see the impact on latency and throughput.  
Let's go gradually to get a sense of what is the impact of each command and parameter.

First, we start by launching our inference model without any tuning, and we observe how the computations are being dispatched on CPU cores (_Left_).

```shell
python3 src/main.py model=bert-base-cased backend.name=pytorch batch_size=1 sequence_length=128
```

Then we specify the core and memory affinity through `numactl` spawning all (_and only_) the **physical** cores (_Right_):
```shell
numactl -C 0-47 -m 0,1 python3 src/main.py model=bert-base-cased backend.name=pytorch batch_size=1 sequence_length=128
```

<img class="centered" alt="htop CPU usage without and with numactl process affinity set" src="assets/19_benchmark_2021_part1/imgs/numa_combined.svg" /> 

> As you can see, without any specific tuning, PyTorch and TensorFlow dispatch the work on a single core, using both physical **and** logical cores.  

Reviewer 2: ```suggestion
As you can see, without any specific tuning, PyTorch and TensorFlow dispatch the work on a single CPU, using both physical **and** logical cores.  
```
Shouldn't be it "CPU" or "processor" here instead of "core" ? This sentence is confusing

<style>
  .centered {
      display: block;
      margin: 0 auto;
  }

  figure {
      text-align: center;
      display: table;
      max-width: 85%; /* demo; set some amount (px or %) if you can */
      margin: 10px auto; /* not needed unless you want centered */
  }
</style>

<div class="blog-metadata">
    <small>Published April 19, 2021.</small>
</div>

<div class="author-card">
    <a href="/mfuntowicz">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1583858935715-5e67c47c100906368940747e.jpeg?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>mfuntowicz</code>
            <span class="fullname">Morgan Funtowicz</span>
        </div>
    </a>

> </div>

Reviewer 3: For consistency, can you add the customary button/link to edit the blogpost on GitHub, like on other blogposts? @mfuntowicz
